[NVFuser] always use fallback if fusion fails

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75983

[NVFuser] always fallback if fusion fails

1) remember when fusions fail; and on subsequent runs, always take the fallback.
2) during the first fallback, cache the Code object.

On autogen-69 from the nvfuser microbenchmarks (https://github.com/pytorch/benchmark/pull/801) this improved performanance as follows:
* Original (always attempt fusion): 25ms
* Always take fallback after first failure: 0.79ms
* Always take fallback + cache Code object: 0.62ms
* Eager: 0.58ms

davidberard98 committed:
[WIP] caching interpreter state or code or something

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] caching interpreter state or code or something"




[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] caching interpreter state or code or something"




[ghstack-poisoned]

davidberard98 committed:
Update on "[NVFuser] always use fallback if fusion fails"


[NVFuser] always fallback if fusion fails

1) remember when fusions fail; and on subsequent runs, always take the fallback.
2) during the first fallback, cache the Code object.

On autogen-69 from the nvfuser microbenchmarks (https://github.com/pytorch/benchmark/pull/801) this improved performanance as follows:
* Original (always attempt fusion): 25ms
* Always take fallback after first failure: 0.79ms
* Always take fallback + cache Code object: 0.62ms
* Eager: 0.58ms

[ghstack-poisoned]

jjsjann123 commented on torch/csrc/jit/codegen/cuda/manager.cpp:
```diff
@@ -52,6 +52,25 @@ namespace cuda {
 
+// Mark string attribute in alias-copy nodes to enable its implementation
+// in the fallback path.
+void enableAliasCopyNodes(const std::shared_ptr<Graph>& graph, Block* block) {
```
> we don't need this any more. I think we can use `revertAliasCopyOps`. Similar refactor should be done infallback graph used by fusion guard. implementation of xxxx_copy ops should be removed.

nitpick, I'll do a clean up afterwards. Do you mind putting a TODO here and copy-paste

davidberard98 replied:
Added the TODO

jjsjann123 commented on torch/csrc/jit/codegen/cuda/manager.cpp:
```diff
@@ -110,6 +129,30 @@ class CudaFusionManager {
 
+  bool hasFallbackCode(int32_t kernel_id) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    return graph_cache_.count(kernel_id);
```
shouldn't this be `fallback_cache_` instead?

davidberard98 replied:
Whoops, thanks for catching that!

davidberard98 committed:
Update on "[NVFuser] always use fallback if fusion fails"


[NVFuser] always fallback if fusion fails

1) remember when fusions fail; and on subsequent runs, always take the fallback.
2) during the first fallback, cache the Code object.

On autogen-69 from the nvfuser microbenchmarks (https://github.com/pytorch/benchmark/pull/801) this improved performanance as follows:
* Original (always attempt fusion): 25ms
* Always take fallback after first failure: 0.79ms
* Always take fallback + cache Code object: 0.62ms
* Eager: 0.58ms

[ghstack-poisoned]

jjsjann123 reviewed (approved):
LGTM.

davidberard98 commented:
@pytorchmergebot merge this

datumbox commented:
@davidberard98 We have started receiving "RuntimeError: required keyword attribute 'cache_id' is undefined" error messages at TorchVision after the landing of this PR. We are not 100% but we might be affected by it. Could you please have a look at: https://github.com/pytorch/vision/issues/5881

davidberard98 commented:
@pytorchmergebot revert this

@datumbox thanks for reporting it, I'll take a look

jjsjann123 commented:
> @pytorchmergebot revert this
> 
> @datumbox thanks for reporting it, I'll take a look

We could have fallback prior to getting a attr::cache_id set
https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L209-L219

Fallback seems to be using that same id.
https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L299

I think we needed to setup the cache_id for fallback in the catch here: https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L226-L231

jjsjann123 commented:
> @davidberard98 We have started receiving "RuntimeError: required keyword attribute 'cache_id' is undefined" error messages at TorchVision after the landing of this PR. We are not 100% but we might be affected by it. Could you please have a look at: [pytorch/vision#5881](https://github.com/pytorch/vision/issues/5881)

Meanwhile, there seems to be a real failure here. Need to take a look
